From b38f3e245fb5239b9749796fdfaaa56384ff8b8d Mon Sep 17 00:00:00 2001
From: Mikhail Lukashov <mlukashov@tabit.pro>
Date: Thu, 23 Dec 2021 17:33:35 +0300
Subject: [PATCH] qubesdb-daemon: fix errors:
 IOCTL_XENIFACE_GNTTAB_UNMAP_FOREIGN_PAGES failed when libvchan_close() called

---
 daemon/db-daemon.c | 26 ++++++++++++++++++++------
 1 file changed, 20 insertions(+), 6 deletions(-)

diff --git a/daemon/db-daemon.c b/daemon/db-daemon.c
index 5474160..d3a4d48 100644
--- a/daemon/db-daemon.c
+++ b/daemon/db-daemon.c
@@ -193,7 +193,7 @@ int mainloop(struct db_daemon_data *d) {
         case 1: {
             // service stopped
             LogInfo("service stopped, exiting");
-            return 1;
+            goto cleanup;
         }
 
         case 2: {
@@ -243,6 +243,21 @@ int mainloop(struct db_daemon_data *d) {
         }
         }
     }
+
+cleanup:
+    if (WaitForSingleObject(pipe_thread, 1000) != WAIT_OBJECT_0)
+    {
+        TerminateThread(pipe_thread, 0);
+        CloseHandle(pipe_thread);
+    }
+    QpsDestroy(d->pipe_server);
+    d->pipe_server = NULL;
+    if (d->vchan)
+    {
+        libvchan_close(d->vchan);
+        d->vchan = NULL;
+    }
+
     return 1;
 }
 
@@ -271,6 +286,7 @@ DWORD WINAPI pipe_thread_client(PVOID param) {
             return 1;
         }
     }
+    return 0;
 }
 
 void client_connected_callback(PIPE_SERVER server, LONGLONG id, PVOID context) {
@@ -338,7 +354,8 @@ int init_server_socket(struct db_daemon_data *d) {
 }
 
 void close_server_socket(struct db_daemon_data *d) {
-    QpsDestroy(d->pipe_server);
+    if (d->pipe_server)
+        QpsDestroy(d->pipe_server);
     d->pipe_server = NULL;
 }
 #endif // WIN32
@@ -686,7 +703,7 @@ int fuzz_main(int argc, char **argv) {
         exit(1);
     }
 
-    memset(&d, 0, sizeof(d));
+    RtlSecureZeroMemory(&d, sizeof(d));
 
     d.remote_domid = atoi(argv[1]);
     if (argc >= 3 && strlen(argv[2]) > 0)
@@ -820,9 +837,6 @@ int fuzz_main(int argc, char **argv) {
     ret = !mainloop(&d);
 #endif /* !WIN32 */
 
-    if (d.vchan)
-        libvchan_close(d.vchan);
-
     close_server_socket(&d);
 
 #ifndef WIN32
-- 
2.34.1.windows.1

